aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -863,6 +893,134 @@ std::tuple<Tensor&,Tensor&> qr_out(Tensor& Q, Tensor& R, const Tensor& self, boo
   return std::tuple<Tensor&, Tensor&>(Q, R);
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ syevd ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+template <typename scalar_t>
+static void apply_syevd(Tensor& w, Tensor& v, bool compute_v, std::string uplo_str, std::vector<int64_t>& infos) {
+#ifndef USE_LAPACK
+  AT_ERROR("syevd: LAPACK library not found in compilation");
+#else
+  using value_t = typename c10::scalar_value_type<scalar_t>::type;
+
+  auto v_data = v.data_ptr<scalar_t>();
+  auto w_data = w.data_ptr<value_t>();
+  auto v_matrix_stride = matrixStride(v);
+  auto w_stride = w.size(-1);
+  auto batch_size = batchCount(v);
+  auto n = v.size(-1);
+  auto lda = std::max(int64_t{1}, n);
+
+  char uplo = uplo_str == "U" ? 'U' : 'L';
+  char jobz = compute_v ? 'V' : 'N';
+
+  int info;
+  // Run once, first to get the optimum work size.
+  // Since we deal with batches of matrices with the same dimensions, doing this outside
+  // the loop saves (batch_size - 1) workspace queries which would provide the same result
+  // and (batch_size - 1) calls to allocate and deallocate workspace using at::empty()
+  int lwork = -1;
+  int lrwork = -1;
+  int liwork = -1;
+  scalar_t work_query;
+  value_t rwork_query;
+  int iwork_query;
+
+  lapackSyevd<scalar_t, value_t>(jobz, uplo, n, v_data, lda, w_data, &work_query, lwork, &rwork_query, lrwork, &iwork_query, liwork, &info);
+  lwork = std::max(1, static_cast<int>(real_impl<scalar_t, value_t>(work_query)));
+  Tensor work = at::empty({lwork}, v.options());
+  liwork = std::max(1, static_cast<int>(iwork_query));
+  Tensor iwork = at::empty({liwork}, at::kInt);
+
+  Tensor rwork;
+  value_t* rwork_data = nullptr;
+  if (isComplexType(at::typeMetaToScalarType(v.dtype()))) {
+    lrwork = std::max(1, static_cast<int>(rwork_query));
+    rwork = at::empty({lrwork}, w.options());
+    rwork_data = rwork.data_ptr<value_t>();
+  }
+
+  for (int64_t i = 0; i < batch_size; i++) {
+    scalar_t* v_working_ptr = &v_data[i * v_matrix_stride];
+    value_t* w_working_ptr = &w_data[i * w_stride];
+    lapackSyevd<scalar_t, value_t>(jobz, uplo, n, v_working_ptr, lda, w_working_ptr, work.data_ptr<scalar_t>(), lwork, rwork_data, lrwork, iwork.data_ptr<int>(), liwork, &info);
+    infos[i] = info;
+    if (info != 0) {
+      return;
```

Comment:
Add a comment here explaining what triggers this return, and what callers should do if this is triggered

Reply:
Follow-up question:

In the future we expect to add `_ex` variants of functions in linear algebra that have additional error information. For example, let's say we have `torch.linalg.foo` that computes the `foo` operation on a non-singular matrix. If the matrix is singular, however, then the math library we're using returns error information on a device tensor explaining that the matrix is singular. 

On the CPU, torch.linalg.foo(cpu_matrix) can throw an error when a singular tensor is encountered. On CUDA, however, translating the device tensor containing the error information requires a cross-device sync, which we try to avoid. torch.linalg.foo(cuda_matrix) should perform and document this sync, however.

torch.linalg.foo_ex, however, will return both device tensors and not perform the error check. This allows the user to check the error themselves, and avoid an immediate cross-device sync on CUDA if they want.

Reply:
The current behavior for linalg (LAPACK/MAGMA based) functions to raise an error if something goes wrong or input doesn't satisfy some requirement therefore all batched CPU functions return early since further computations will be wasted anyway. In the future, this early return should be removed.

What does `_ex` suffix mean?
I agree that this kind of change will improve performance. I think the default behavior should be the same on CPU and CUDA, so CPU shouldn't raise an error, even though it's cheap to check for it, if CUDA doesn't.
What do you think about instead of introducing new functions with `_ex` suffix in Python interface modify existing ones to accept optional `info=` argument, similar to `out=`, that will be filled with the returned LAPACK/MAGMA/cuSOLVER error codes for the user to check if he wants to.
```python
eigenvalues, eigenvectors = torch.linalg.eigh(input, info=torch.empty(batch_size))
# instead of 
eigenvalues, eigenvectors, info = torch.linalg.eigh_ex(input)
```

Reply:
> The current behavior for linalg (LAPACK/MAGMA based) functions to raise an error if something goes wrong or input doesn't satisfy some requirement therefore all batched CPU functions return early since further computations will be wasted anyway. In the future, this early return should be removed.

Sounds good for now.

 
> What does `_ex` suffix mean?

"Extra."

> I agree that this kind of change will improve performance. I think the default behavior should be the same on CPU and CUDA, so CPU shouldn't raise an error, even though it's cheap to check for it, if CUDA doesn't.

Agreed. Both the CPU and CUDA ex variants will not perform their own error checking and return the info tensor for callers to evaluate.

> What do you think about instead of introducing new functions with `_ex` suffix in Python interface modify existing ones to accept optional `info=` argument, similar to `out=`, that will be filled with the returned LAPACK/MAGMA/cuSOLVER error codes for the user to check if he wants to.

This is a really good idea and we did consider an approach like this, but there are technical issues with this approach that make it prohibitive.